### ISSUE TITLE
operator: allow restarting or updating without killing driver installations

### ISSUE THREAD

**Issue by pohly** [intel/pmem-csi#579]
Steps:
- install operator
- deploy driver
- re-install or update operator

Expected behavior:
- the driver deployment is accessible while the operator is not running
- when the operator restarts, it reconciles the current deployment object with the running driver

Actual behavior:
- removing the operator *pod* deletes everything

**Comment by avalluri** [comment]
I tried the steps as below:

```
$ TEST_OPERATOR_NAMESPACE=test ./test/start-operator.sh 
Deploying '/home/avalluri/work/pmem-csi/operator/deploy/operator.yaml'...
serviceaccount/pmem-csi-operator created
role.rbac.authorization.k8s.io/pmem-csi-operator created
clusterrole.rbac.authorization.k8s.io/pmem-csi-operator created
rolebinding.rbac.authorization.k8s.io/pmem-csi-operator created
clusterrolebinding.rbac.authorization.k8s.io/pmem-csi-operator created
deployment.apps/pmem-csi-operator created
PMEM-CSI operator is running. To try out deploying the pmem-csi driver:
    cat operator/examples/pmem-csi.intel.com_v1alpha1_deployment_cr.yaml | /home/avalluri/work/pmem-csi/_work/pmem-govm/ssh.0 kubectl create -f -
cleaning up temp folder: /tmp/tmp.OiZkFmCNp0

$ kubectl get po -n test
NAME                                 READY   STATUS    RESTARTS   AGE
pmem-csi-operator-55b4bbd66d-qvb2t   1/1     Running   0          15s

$ kubectl create -f operator/examples/pmem-csi.intel.com_v1alpha1_deployment_cr.yaml 
deployment.pmem-csi.intel.com/example-deployment created

$ kubectl get po -n test
NAME                                 READY   STATUS              RESTARTS   AGE
example-deployment-controller-0      2/2     Running   0          47s
example-deployment-node-d5pv4        2/2     Running   0          47s
example-deployment-node-qzwbg        2/2     Running   0          47s
example-deployment-node-wq7rg        2/2     Running   0          47s
pmem-csi-operator-55b4bbd66d-qvb2t   1/1     Running   0          116s

$ kubectl delete deployment --all -n test
deployment.apps "pmem-csi-operator" deleted

$ kubectl get po -n test
NAME                              READY   STATUS    RESTARTS   AGE
example-deployment-controller-0   2/2     Running   0          2m13s
example-deployment-node-d5pv4     2/2     Running   0          2m13s
example-deployment-node-qzwbg     2/2     Running   0          2m13s
example-deployment-node-wq7rg     2/2     Running   0          2m13s

```

**Comment by pohly** [comment]
Can you do `kubectl describe deployments.pmem-csi.intel.com` while the operator is gone?

**Comment by pohly** [comment]
Doesn't work for me:
```
pohly@pohly-desktop:/nvme/gopath/src/github.com/intel/pmem-csi$ kubectl delete deployment --all -n test
deployment.apps "pmem-csi-operator" deleted
pohly@pohly-desktop:/nvme/gopath/src/github.com/intel/pmem-csi$ kubectl get po -n test
NAME                                READY   STATUS        RESTARTS   AGE
example-deployment-controller-0     0/2     Terminating   0          14s
example-deployment-node-7dfw2       0/2     Terminating   0          14s
example-deployment-node-dwvdj       2/2     Terminating   0          14s
example-deployment-node-rtdxz       0/2     Terminating   0          14s
pmem-csi-operator-d8448dc5b-tpl88   0/1     Terminating   0          34s
```

Can you write an E2E test for this? Then we have a reproducible way of testing this feature.

**Comment by pohly** [comment]
It's possible that I broke that when replacing several different `ObjectMeta` instantiations with a single function. I'll check that next Tuesday after the Easter vacation.

A test for this is still needed, to catch such regressions.

**Comment by pohly** [comment]
I would implement this like this:
- change `deploy.EnsureDeployment` so that it always goes through the install steps (which are resp. must be idempotent), i.e. it'll fix an incomplete deployment - that is useful by itself
- in the test:
  - identify some driver object(s)
  - find the operator deployment and remove it, with waiting to ensure that the pod is really gone
  - call  `deploy.EnsureDeployment` to restart the operator
  - check that the driver objects haven't changed; a very strict check is to compare the generation field

This should built on top of the deploy package from https://github.com/intel/pmem-csi/pull/575 to avoid code conflicts.

**Comment by pohly** [comment]
> I tried the steps as below: ...

Have you figured out why it worked for you? Apparently you were able to reproduce it later (https://github.com/intel/pmem-csi/pull/584).

**Comment by avalluri** [comment]
> I tried the steps as below: ...
> 
> Have you figured out why it worked for you? Apparently you were able to reproduce it later (#584).

Yes, @pohly I figured out the issue, It was with the way I stop the operator(``kubectl delete -f operator/deploy/operator.yaml`) It was deleting first clsterrole/binding and then operator deployment. So delete CRD by operator failed with a lack of permissions.

Now I fixed this in #584

**Comment by pohly** [comment]
Fixed, and the test should cover this.
